cpp/ql/src/Likely Bugs/Memory Management/PointerOverflowCheck.ql

```diff
@@ -0,0 +1,36 @@
+/**
+ * @name Dangerous pointer out-of-range check
```

Comment:
So then we'll have something like `p >= INTPTR_MAX - n`, where `INTPTR_MAX - n` is guaranteed not to overflow provided that `n > 0`.

Reply:
I'm confused. What do those pointer calculations have to do with the name of the query?

In any case, we can't calculate with pointers like they're integers, even if we cast them to `intptr_t`. There's no guarantee that if `p1 < p2` then `(intptr_t)p1 < (intptr_t)p2`. The compiler is in principle allowed to mirror the bits when converting between pointers and integers. (I'm sure it will be fine on x86, though).

Reply:
Now it's my turn to be confused. :-) Pointer difference is a feature present in the C/C++ language, and it imposes a total linear order among pointer values.  Furthermore, C99 introduces INTPTR_MAX, which is the "highest" value in said total order.  Perhaps the inequality above should have been written as `n > (char *)INTPTR_MAX - (char *)p`, which is a mouthful, but still good advice, I believe.